cpg-concepts/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/concepts/LoadPersistedConcepts.kt

```diff
@@ -0,0 +1,318 @@
+                val endLine =
+                    region.groups["endLine"]?.value?.toIntOrNull()
+                        ?: throw IllegalArgumentException("Invalid or missing endLine in region.")
+                val endColumn =
+                    region.groups["endColumn"]?.value?.toIntOrNull()
+                        ?: throw IllegalArgumentException("Invalid or missing endColumn in region.")
+
+                PhysicalLocation(
+                    uri = URI(location.file),
+                    region = Region(startLine, startColumn, endLine, endColumn),
+                )
+            } catch (ex: Exception) {
+                log.error("Failed to parse the location: ${location.region}: ${ex.message}", ex)
+                return emptyList()
+            }
+
+        // find the matching node
+        return translationResult.getNodesByRegion(location = loc, clsName = location.type).also {
+            nodes ->
+            if (nodes.size != 1) {
+                logger.warn(
+                    "Found ${nodes.size} nodes for location $loc. Expected exactly one node."
+                )
+            }
+        }
+    }
+
+    /**
+     * This is a helper function to add a [Concept] node to the [underlyingNode] as described in the
+     * provided [ConceptEntry].
+     *
+     * @param underlyingNode The node to which the concept will be added.
+     * @param concept The [ConceptEntry] containing the concept information.
+     */
+    private fun addConcept(underlyingNode: Node, concept: ConceptEntry) {
+        logger.debug("Adding concept {} to node {}.", concept, underlyingNode)
+        underlyingNode.conceptBuildHelper(
+            name = concept.name,
+            underlyingNode = underlyingNode,
+            constructorArguments =
+                concept.constructorArguments?.associate { arg -> arg.name to arg.value }
+                    ?: emptyMap(),
+            connectDFGUnderlyingNodeToConcept =
+                concept.dfg?.fromThisNodeToConcept ?: false, // TODO: this `?: false` is not nice
+            connectDFGConceptToUnderlyingNode =
+                concept.dfg?.fromConceptToThisNode ?: false, // TODO: this `?: false` is not nice
+        )
+    }
+
+    /** The root node of our YAML/JSON structure. It contains a list of [ConceptRootEntry]s. */
+    private data class PersistedConceptsEntry(val concepts: List<ConceptRootEntry>?)
+
+    /**
+     * This class represents a single entry in the YAML/JSON file. It contains the concept and
+     * optionally a location and/or a signature entry. The pass itself enforces that exactly one of
+     * location or signature is set.
+     *
+     * @param concept The [ConceptEntry] containing the concept information.
+     * @param location The [LocationEntry] containing the [PhysicalLocation] information to match
+     *   against.
+     * @param signature The [SignatureEntry] containing the signature of nodes to match.
+     */
+    private data class ConceptRootEntry(
```

Comment:
Hmm needs a better name. Maybe `PersistedConceptEntry` for this and `PersistedConceptEntryList` or `PersistedConcepts` for the outer one? (in line 232)